elasticsearch-transport/lib/elasticsearch/transport/transport/http/manticore/pool.rb

```diff
@@ -0,0 +1,201 @@
+                  # If no exception was raised it must have succeeded!
+                  logger.warn("Resurrected connection to dead ES instance at #{url}")
+                  @state_mutex.synchronize { meta[:dead] = false }
+                rescue ::Elasticsearch::Transport::Transport::HostUnreachableError => e
+                  logger.debug("Attempted to resurrect connection to dead ES instance at #{url}, got an error [#{e.class}] #{e.message}")
+                end
+              end
+            end
+
+            def stop_resurrectionist
+              @state_mutex.synchronize { @stopping = true }
+              @resurrectionist.join # Wait for thread to stop
+            end
+
+            def perform_request(method, path, params={}, body=nil)
+              with_connection do |url|
+                [url, perform_request_to_url(url, method, path, params, body)]
+              end
+            end
+
+            def perform_request_to_url(url, method, path, params={}, body=nil)
+              raise ArgumentError, "No URL specified" unless url
+
+              @adapter.perform_request(url, method, path, params, body)
+            rescue ::Manticore::Timeout,::Manticore::SocketException, ::Manticore::ClientProtocolException, ::Manticore::ResolutionFailure => e
+              logger.error "[#{e.class}] #{e.message} #{url}" if logger
+              raise ::Elasticsearch::Transport::Transport::HostUnreachableError.new(e, url), "Could not reach host #{e.class}: #{e.message}"
+            end
+
+            def update_urls(new_urls)
+              # Normalize URLs
+              new_urls = new_urls.map {|u| u.is_a?(URI) ? u : URI.parse(u) }
+
+              @state_mutex.synchronize do
+                # Add new connections
+                new_urls.each do |url|
+                  # URI objects don't have real hash equality! So, since this isn't perf sensitive we do a linear scan
+                  unless @url_info.keys.include?(url)
+                    logger.info("Elasticsearch pool adding node @ URL #{url}") if logger
+                    add_url(url)
+                  end
+                end
+
+                # Delete connections not in the new list
+                @url_info.each do |url,_|
+                  unless new_urls.include?(url)
+                    logger.info("Elasticsearch pool removing node @ URL #{url}") if logger
+                    remove_url(url)
+                  end
+                end
+              end
+            end
+
+            def add_url(url)
+              @url_info[url] ||= empty_url_meta
+            end
+
+            def remove_url(url)
+              @url_info.delete(url)
+            end
+
+            def empty_url_meta
+              {
+                :open => 0,
+                :dead => false
+              }
+            end
+
+            def with_connection
+              url, url_meta = get_connection
+
+              raise Error, "No Available connections" unless url
```

Comment:
should `get_connection` really throw an exception if there are no connections available? I'd somehow expect this to be blocking or, at least, have that option (similar to get/try)